client/agent_endpoint.go

```diff
@@ -233,3 +233,124 @@ func (a *Agent) Host(args *structs.HostDataRequest, reply *structs.HostDataRespo
 	reply.HostData = data
 	return nil
 }
+func (a *Agent) monitorExport(conn io.ReadWriteCloser) {
+	defer conn.Close()
+
+	// Decode arguments
+	var args cstructs.MonitorExportRequest
+
+	decoder := codec.NewDecoder(conn, structs.MsgpackHandle)
+	encoder := codec.NewEncoder(conn, structs.MsgpackHandle)
+
+	if err := decoder.Decode(&args); err != nil {
+		handleStreamResultError(err, pointer.Of(int64(500)), encoder)
+		return
+	}
+
+	// Check acl
+	if aclObj, err := a.c.ResolveToken(args.AuthToken); err != nil {
+		handleStreamResultError(err, pointer.Of(int64(403)), encoder)
+		return
+	} else if !aclObj.AllowAgentRead() {
+		handleStreamResultError(structs.ErrPermissionDenied, pointer.Of(int64(403)), encoder)
+		return
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	mon := monitor.New(512, a.c.logger, &log.LoggerOptions{})
+	if args.MockMonitor != nil {
+		mon = args.MockMonitor
+	}
+	frames := make(chan *sframer.StreamFrame, streamFramesBuffer)
+	errCh := make(chan error)
+	var buf bytes.Buffer
+	frameCodec := codec.NewEncoder(&buf, structs.JsonHandle)
+
+	framer := sframer.NewStreamFramer(frames, 1*time.Second, 200*time.Millisecond, 1024)
+	framer.Run()
+
+	defer framer.Destroy()
+	// goroutine to detect remote side closing
+	go func() {
+		if _, err := conn.Read(nil); err != nil {
+			// One end of the pipe explicitly closed, exit
+			cancel()
+			return
+		}
+		<-ctx.Done()
+	}()
+
+	opts := monitor.MonitorExportOpts{
+		LogSince:     args.LogSince,
+		ServiceName:  args.ServiceName,
+		NomadLogPath: args.NomadLogPath,
+		OnDisk:       args.OnDisk,
+		Follow:       args.Follow,
+	}
+
+	logCh := mon.MonitorExport(opts)
+
+	initialOffset := int64(0)
+	var (
+		eofCancelCh chan error
+		eofCancel   bool
+	)
+	eofCancel = !opts.Follow
+	// receive logs and build frames
+	streamReader := cstructs.NewStreamReader(logCh)
+	go func() {
+		defer framer.Destroy()
+
+		if err := streamReader.StreamFixed(ctx, initialOffset, "", 0, framer, eofCancelCh, eofCancel); err != nil {
+			select {
+			case errCh <- err:
+			case <-ctx.Done():
+			}
+		}
+	}()
+	var streamErr error
```

Comment:
This chunk seems to be the primary difference between this and the `Agent.monitor` RPC, which as you'll guess from my other comments suggests to me they could share a lot of code. It seems like we're repeating all the stream handling code in the RPC handlers when we could reuse that if we have 2 different monitor implementations with the same interface (i.e. the `Start`/`End` interface), instead of introducing a new method into the existing struct.